File: mines_app/lib/sidebar.dart
import 'package:flutter/material.dart';
import 'package:mines_app/clickableObject.dart';

class MenuButton extends ClickableObject {
  final VoidCallback? onRightMouseClick;

  MenuButton({
    required super.x,
    required super.y,
    required super.width,
    required super.height,
    required super.defaultColor,
    required super.text,
    VoidCallback? onLeftMouseClick,
    this.onRightMouseClick,
  }) : super(
          onLeftMouseClick: onLeftMouseClick ??
              () => print(1), // Default action for left click
        );

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onLeftMouseClick, // Left mouse click (tap)
      onSecondaryTap: onRightMouseClick, // Right mouse click equivalent
      child: Container(
        width: width,
        height: height,
        color: defaultColor,
        child: Center(
          child: Text(
            text,
            style: const TextStyle(
              fontSize: 36.0,
              color: Colors.black,
            ),
          ),
        ),
      ),
    );
  }
}

class Sidebar extends StatelessWidget {
  final VoidCallback resetMinefield;
  final VoidCallback exitGame;

  const Sidebar({
    super.key,
    required this.resetMinefield,
    required this.exitGame,
  });

  @override
  Widget build(BuildContext context) {
    double buttonWidth = 150;
    double buttonHeight = 50;
    double buttonMargin = 20;

    return Container(
      width: buttonWidth + 40, // Add some margin for the container
      padding: const EdgeInsets.all(10), // Padding around the buttons
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          MenuButton(
            x: 0, // Position will be managed by the Column
            y: 0, // Position will be managed by the Column
            width: buttonWidth,
            height: buttonHeight,
            defaultColor: Colors.white,
            text: "Reset",
            onLeftMouseClick: resetMinefield,
          ),
          SizedBox(height: buttonMargin), // Space between buttons
          MenuButton(
            x: 0, // Position will be managed by the Column
            y: 0, // Position will be managed by the Column
            width: buttonWidth,
            height: buttonHeight,
            defaultColor: Colors.white,
            text: "Exit",
            onLeftMouseClick: exitGame,
          ),
        ],
      ),
    );
  }
}

File: pygame/sidebar.py
import pygame
from pygame.locals import *
import random
import types

from clickableObject import ClickableObject, RED, WHITE


class MenuButton(ClickableObject):
    def __init__(self, x, y, width, height, default_color, text, on_left = (lambda:print(1)), on_right=(lambda:None)):
        super().__init__(x, y, width, height, default_color, text)
        self.on_left = on_left
        self.on_right = on_right
    def on_leftmouseclick(self):
        self.on_left()
    def on_rightmouseclick(self):
        self.on_right()

    

class Sidebar():
    def __init__(self, game):
        self.game = game
        def reset_on_leftmouseclick():
            game.reset_minefield()
        def exit_on_leftmouseclick():
            print("yes")
            exit(0)

        button_x_offset = 60 
        button_y_offset = 10
        button_width = 15 
        button_height = 40
        button_margin = 20 
        self.resetbutton = MenuButton(button_x_offset,
                                    button_y_offset,
                                    button_width,
                                    button_height,
                                    WHITE, "Reset", on_left=reset_on_leftmouseclick)
        self.exitbutton = MenuButton(button_x_offset,
                                    button_y_offset + button_height + button_margin,
                                    button_width,
                                    button_height,
                                    WHITE, "Exit", on_left=exit_on_leftmouseclick)

        self.buttons = [self.resetbutton, self.exitbutton]
    
    def event(self, event):
        if event.type == MOUSEBUTTONDOWN:
            if event.button == 1:  # Left mouse button clicked
                pos = pygame.mouse.get_pos()
                for button in self.buttons:
                    if button.rect.collidepoint(pos):
                        button.on_leftmouseclick()

            elif event.button == 3:  # Right mouse button clicked
                pos = pygame.mouse.get_pos()
                for button in self.buttons:
                    if button.rect.collidepoint(pos):
                        button.on_rightmouseclick()
